src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -276,35 +281,80 @@ export class ObjectExplorerService implements IObjectExplorerService {
+					reject(new Error(nls.localize(
+						'objectExplorerMissingConnectionForSession',
+						'Timeout waiting for session {0} to be created for connection "{0}"', sessionId, connection.title
+					)));
 				}
-				// Send on session created about the session to all node providers so they can prepare for node expansion
-				let nodeProviders = this._nodeProviders[connection!.providerName];
-				if (nodeProviders) {
-					const promises = nodeProviders.map(p => p.handleSessionOpen(session));
-					await Promise.all(promises);
+
+				cleanup();
+			}
+			const timeout = setTimeout(onTimeout, this.getObjectExplorerTimeout() * 1000);
+			const createNewSessionListener = this._onCreateNewSession.event((response) => {
+				checkSessionAndConnection();
+			});
+			const checkSessionAndConnection = () => {
+				/**
+				 * Sometimes ads recieves handleSessionCreated from providers before the createNewSession response is recieved.
+				 * We need to wait while the createNewSession response is recieved and the session map contains the session before we can continue.
+				 */
+				if (this._sessions[sessionId]) {
+					connection = this._sessions[sessionId].connection;
+					/**
+					 * In certain cases, when we try to connect to a previously connected server, we may encounter a situation where the session is present in this._sessions,
+						 * but the connection is outdated. This happens when the handleSessionCreated is recieved before the createNewSession response is recieved.
+					 * We are using a map to keep track of connections which are actually waiting for a session to be created. And we will continue only if we get createNewSession
+					 * reponse for those connections.
+					 */
+					if (connection && this._connectionsWaitingForSession.get(connection.id) === false) {
+						resolve();
+						cleanup();
+					}
 				}
-			} catch (error) {
-				this.logService.warn(`cannot handle the session ${session.sessionId} in all nodeProviders`);
-			} finally {
-				this.sendUpdateNodeEvent(connection!, errorMessage);
 			}
+			checkSessionAndConnection();
+		});
+
+		try {
+			if (session.success && session.rootNode) {
+				let server = this.toTreeNode(session.rootNode, undefined);
+				server.connection = connection;
+				server.session = session;
+				this._activeObjectExplorerNodes[connection!.id] = server;
+			}
+			else {
+				errorMessage = session && session.errorMessage ? session.errorMessage : errSessionCreateFailed;
+				this.logService.error(errorMessage);
+			}
+			// Send on session created about the session to all node providers so they can prepare for node expansion
+			let nodeProviders = this._nodeProviders[connection!.providerName];
+			if (nodeProviders) {
+				const promises = nodeProviders.map(p => p.handleSessionOpen(session));
+				await Promise.all(promises);
+			}
+		} catch (error) {
+			this.logService.warn(`cannot handle the session ${sessionId} in all nodeProviders`);
```

Comment:
the actual error message is ignored here, don't you need it to troubleshoot?

Reply:
Good catch. Updated the error message to include the error as well.